File: macos/TaskStrips/Sources/TaskStrips/Sync/BoardSyncRunner.swift
import Foundation
import SwiftData

/// The sync, wired to this machine.
///
/// Mirrors BoardSyncRunner.kt. BoardSync decides what should happen; this reads the objects to hand
/// it and writes back what it decided. Everything with judgement in it stays over there, tested;
/// what is here is the store and filesystem work that only running it can prove.
///
/// Credentials travel without their passwords unless `passphrase` is given, which is the rule the
/// backup already follows: the Keychain's copy is this machine's alone, and the alternative to a
/// passphrase is writing secrets in the clear.
@MainActor
struct BoardSyncRunner {
    var context: ModelContext
    var transport: BoardTransport
    var attachments: AttachmentStore = .shared
    var sketches: SketchStore = .shared
    var credentials: CredentialStore = .shared
    var passphrase: String?

    func run(hasSyncedBefore: Bool, adoptsOnFirstSync: Bool) async throws -> BoardSyncOutcome {
        let tasks = try context.fetch(FetchDescriptor<TaskItem>())
        let reminders = try context.fetch(FetchDescriptor<Reminder>())
        let items = try context.fetch(FetchDescriptor<StorageItem>())
        let logins = try context.fetch(FetchDescriptor<Credential>())
        let sketchIDs = sketches.allIDsForSync()

        // Hashed once, up front. Every later question — what to upload, what a strip's attachments
        // are called, where an arriving hash already lives — is answered from this one pass rather
        // than by reading the same files again.
        var pathsByHash: [String: URL] = [:]
        // The store can turn a relative path into a URL but not back again, so both are kept. A
        // record says nothing about where this machine puts a file; the relative path is how this
        // machine names it, and an attachment row needs that rather than an absolute URL.
        var relativeByHash: [String: String] = [:]
        func hashed(_ url: URL, relative: String? = nil) -> String? {
            guard let hash = try? SyncFileStore.hash(contentsOf: url) else { return nil }
            if pathsByHash[hash] == nil { pathsByHash[hash] = url }
            if let relative, relativeByHash[hash] == nil { relativeByHash[hash] = relative }
            return hash
        }

        let syncIDByLocalID = Dictionary(uniqueKeysWithValues: tasks.map { ($0.id, $0.id.uuidString) })
        let sketchSyncIDByFolder = Dictionary(uniqueKeysWithValues: sketchIDs.map { ($0, sketches.syncID(of: $0)) })
        let folderBySketchSyncID = Dictionary(uniqueKeysWithValues: sketchSyncIDByFolder.map { ($0.value, $0.key) })

        let taskRecords = tasks.map { task in
            BoardRecords.record(
                for: task,
                attachments: task.attachments.compactMap { attachment in
                    let url = attachments.url(for: attachment)
                    guard let hash = hashed(url, relative: attachment.path) else { return nil }
                    return SyncAttachment(hash: hash, name: attachment.name, kind: attachment.kind.rawValue)
                },
                syncIDOfTask: { syncIDByLocalID[$0] },
                syncIDOfSketch: { sketchSyncIDByFolder[$0] }
            )
        }

        let local = BoardSnapshot(
            tasks: taskRecords,
            reminders: reminders.map(BoardRecords.record(for:)),
            storage: items.map { item in
                BoardRecords.record(
                    for: item,
                    hash: hashed(attachments.url(forRelativePath: item.path), relative: item.path) ?? ""
                )
            },
            credentials: logins.map { BoardRecords.record(for: $0, password: portablePassword(for: $0)) },
            sketches: sketchIDs.map { id in
                SyncSketchRecord(
                    id: sketchSyncIDByFolder[id] ?? sketches.syncID(of: id),
                    updatedAt: sketches.isDeleted(id)
                        ? sketches.deletedAt(of: id)
                        : BoardRecords.millis(sketches.lastModified(of: id)),
                    isDeleted: sketches.isDeleted(id),
                    name: sketches.name(of: id) ?? "",
                    pages: sketches.pages(of: id).compactMap { hashed($0) },
                    createdAt: BoardRecords.millis(sketches.createdAt(of: id))
                )
            }
        )

        var landed: [String: (URL, String)] = [:]
        let outcome = try await BoardSync(transport: transport).run(
            local: local,
            localHashes: Set(pathsByHash.keys),
            hasSyncedBefore: hasSyncedBefore,
            adoptsOnFirstSync: adoptsOnFirstSync,
            readFile: { hash in
                guard let url = pathsByHash[hash] else { throw CocoaError(.fileNoSuchFile) }
                return try Data(contentsOf: url)
            },
            writeFile: { hash, data in
                // Named by the hash rather than by anything the record said. A file's own name is
                // the record's business and can differ between devices; where the bytes sit is
                // this machine's, and one name per content keeps a second copy from ever landing.
                let relative = "sync/\(hash)"
                try attachments.write(data, toRelativePath: relative)
                landed[hash] = (attachments.url(forRelativePath: relative), relative)
            }
        )

        for (hash, arrival) in landed {
            pathsByHash[hash] = arrival.0
            relativeByHash[hash] = arrival.1
        }
        apply(local: local, outcome: outcome, pathsByHash: pathsByHash,
              relativeByHash: relativeByHash, folderBySketchSyncID: folderBySketchSyncID)
        return outcome
    }

    /// Only what actually differs is written — a sync that changed nothing must not touch a single
    /// object, or every sync wakes every view watching the store.
    private func apply(
        local: BoardSnapshot,
        outcome: BoardSyncOutcome,
        pathsByHash: [String: URL],
        relativeByHash: [String: String],
        folderBySketchSyncID: [String: String]
    ) {
        applyTasks(local: local, outcome: outcome, relativeByHash: relativeByHash,
                   folderBySketchSyncID: folderBySketchSyncID)
        applyReminders(local: local, outcome: outcome)
        applyStorage(local: local, outcome: outcome, relativeByHash: relativeByHash)
        applyCredentials(local: local, outcome: outcome)
        applySketches(local: local, outcome: outcome, pathsByHash: pathsByHash,
                      folderBySketchSyncID: folderBySketchSyncID)
    }

    private func applyTasks(
        local: BoardSnapshot,
        outcome: BoardSyncOutcome,
        relativeByHash: [String: String],
        folderBySketchSyncID: [String: String]
    ) {
        let existing = Dictionary(
            (try? context.fetch(FetchDescriptor<TaskItem>()))?.map { ($0.id.uuidString, $0) } ?? [],
            uniquingKeysWith: { first, _ in first }
        )
        let plan = SyncBoardPlan.plan(
            localByID: Dictionary(uniqueKeysWithValues: local.tasks.map { ($0.id, $0) }),
            merged: outcome.merged.tasks,
            id: { $0.id },
            isDeleted: { $0.isDeleted }
        )

        for record in plan.insert + plan.update {
            let task: TaskItem
            if let here = existing[record.id] {
                task = here
            } else {
                guard let made = BoardRecords.newTask(from: record) else { continue }
                context.insert(made)
                task = made
            }
            BoardRecords.apply(
                record, to: task,
                localIDOfSyncID: { UUID(uuidString: $0) },
                localSketchIDOf: { folderBySketchSyncID[$0] }
            )
            // Only the attachments whose bytes are actually here. One that hasn't landed is left
            // out rather than pointed at nothing; the next sync adds it once the file arrives.
            task.attachments = record.attachments.compactMap { attachment -> TaskAttachment? in
                guard let relative = relativeByHash[attachment.hash] else { return nil }
                return TaskAttachment(
                    kind: AttachmentKind(rawValue: attachment.kind) ?? .document,
                    path: relative,
                    name: attachment.name
                )
            }
        }
        for id in plan.delete {
            // Written as a tombstone, not removed: this machine has to be able to tell the *next*
            // device about the delete too.
            existing[id].map { $0.isTombstoned = true }
        }
    }

    private func applyReminders(local: BoardSnapshot, outcome: BoardSyncOutcome) {
        let existing = Dictionary(
            (try? context.fetch(FetchDescriptor<Reminder>()))?.map { ($0.id.uuidString, $0) } ?? [],
            uniquingKeysWith: { first, _ in first }
        )
        let plan = SyncBoardPlan.plan(
            localByID: Dictionary(uniqueKeysWithValues: local.reminders.map { ($0.id, $0) }),
            merged: outcome.merged.reminders,
            id: { $0.id },
            isDeleted: { $0.isDeleted }
        )
        for record in plan.insert + plan.update {
            if let here = existing[record.id] {
                BoardRecords.apply(record, to: here)
            } else if let made = BoardRecords.newReminder(from: record) {
                context.insert(made)
                BoardRecords.apply(record, to: made)
            }
        }
        for id in plan.delete { existing[id].map { $0.isTombstoned = true } }
    }

    /// A library item is only worth a row once its bytes are somewhere: one whose file hasn't
    /// landed is skipped rather than written with a path that opens nothing — written, the row
    /// would look up to date and no later sync would fix it.
    private func applyStorage(local: BoardSnapshot, outcome: BoardSyncOutcome, relativeByHash: [String: String]) {
        let existing = Dictionary(
            (try? context.fetch(FetchDescriptor<StorageItem>()))?.map { ($0.id.uuidString, $0) } ?? [],
            uniquingKeysWith: { first, _ in first }
        )
        let plan = SyncBoardPlan.plan(
            localByID: Dictionary(uniqueKeysWithValues: local.storage.map { ($0.id, $0) }),
            merged: outcome.merged.storage,
            id: { $0.id },
            isDeleted: { $0.isDeleted }
        )
        for record in plan.insert + plan.update {
            if let here = existing[record.id] {
                BoardRecords.apply(record, to: here)
            } else {
                guard let id = UUID(uuidString: record.id),
                      let relative = relativeByHash[record.hash]
                else { continue }
                let item = StorageItem(
                    name: record.name,
                    path: relative,
                    type: StorageItemType(rawValue: record.type) ?? .document,
                    id: id,
                    createdAt: BoardRecords.date(record.createdAt)
                )
                context.insert(item)
                BoardRecords.apply(record, to: item)
            }
        }
        for id in plan.delete { existing[id].map { $0.isTombstoned = true } }
    }

    /// The password crosses under the user's passphrase and is kept here in the Keychain, so it is
    /// decrypted and re-stored on the way in. With no passphrase, or a record carrying no secret,
    /// whatever is already in the Keychain is left alone: losing a password to a sync that couldn't
    /// read it would be losing data, and keeping the old one never is.
    private func applyCredentials(local: BoardSnapshot, outcome: BoardSyncOutcome) {
        let existing = Dictionary(
            (try? context.fetch(FetchDescriptor<Credential>()))?.map { ($0.id.uuidString, $0) } ?? [],
            uniquingKeysWith: { first, _ in first }
        )
        let plan = SyncBoardPlan.plan(
            localByID: Dictionary(uniqueKeysWithValues: local.credentials.map { ($0.id, $0) }),
            merged: outcome.merged.credentials,
            id: { $0.id },
            isDeleted: { $0.isDeleted }
        )
        for record in plan.insert + plan.update {
            let credential: Credential
            if let here = existing[record.id] {
                credential = here
            } else {
                guard let id = UUID(uuidString: record.id) else { continue }
                let made = Credential(title: record.title, id: id,
                                      createdAt: BoardRecords.date(record.createdAt))
                context.insert(made)
                credential = made
            }
            BoardRecords.apply(record, to: credential)
            if let secret = arrivingPassword(record) {
                _ = credentials.setPassword(secret, for: credential.id)
            }
        }
        for id in plan.delete { existing[id].map { $0.isTombstoned = true } }
    }

    /// Pages are written in the record's order rather than the order they arrived, because page
    /// numbering is a local detail and the record is what says which page is which. A note whose
    /// pages haven't all landed is left alone entirely — half a drawing is worse than yesterday's
    /// whole one, and the next sync writes it properly once the rest is here.
    private func applySketches(
        local: BoardSnapshot,
        outcome: BoardSyncOutcome,
        pathsByHash: [String: URL],
        folderBySketchSyncID: [String: String]
    ) {
        let here = Dictionary(uniqueKeysWithValues: local.sketches.map { ($0.id, $0) })
        for record in outcome.merged.sketches where record != here[record.id] {
            let folder = folderBySketchSyncID[record.id] ?? "note_\(record.createdAt)"

            if record.isDeleted {
                if !sketches.pages(of: folder).isEmpty { sketches.deleteNote(folder) }
                sketches.setSyncID(record.id, of: folder)
                continue
            }

            let sources = record.pages.map { pathsByHash[$0] }
            guard !sources.contains(where: { $0 == nil }) else { continue }

            sketches.setSyncID(record.id, of: folder)
            for page in sketches.pages(of: folder) { sketches.deletePage(page) }
            for (index, source) in sources.compactMap({ $0 }).enumerated() {
                guard let data = try? Data(contentsOf: source) else { continue }
                try? sketches.write(data, to: sketches.folder(of: folder).appending(path: "page\(index + 1).png"))
            }
            if !record.name.isEmpty { sketches.setName(record.name, of: folder) }
        }
    }

    private func portablePassword(for credential: Credential) -> BoardRecords.PortablePassword? {
        guard let passphrase, let plain = credentials.password(for: credential.id), !plain.isEmpty
        else { return nil }
        guard let sealed = BackupCrypto.encrypt(plain, passphrase: passphrase) else { return nil }
        return BoardRecords.PortablePassword(salt: sealed.salt, iv: sealed.iv, cipher: sealed.cipher)
    }

    private func arrivingPassword(_ record: SyncCredentialRecord) -> String? {
        guard let passphrase, record.hasPassword else { return nil }
        return BackupCrypto.decrypt(
            BackupCrypto.Encrypted(
                salt: record.passwordSalt ?? "",
                iv: record.passwordIv ?? "",
                cipher: record.passwordCipher ?? ""
            ),
            passphrase: passphrase
        )
    }
}

File: app/src/main/java/com/saud/taskstrip/sync/BoardSyncRunner.kt
package com.saud.taskstrip.sync

import android.content.Context
import com.saud.taskstrip.data.AppDatabase
import com.saud.taskstrip.media.MediaStorage
import com.saud.taskstrip.media.SketchStorage
import java.io.File
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * The sync, wired to this device.
 *
 * [BoardSync] decides what should happen; this reads the rows to hand it and writes back what it
 * decided. The split is deliberate — everything with judgement in it is over there and tested, and
 * what is here is the database and filesystem work that only running it can prove.
 *
 * Credentials travel without their passwords unless [passphrase] is given, which is the same rule
 * the backup follows and for the same reason: the stored password is keystore-tied, dead on arrival
 * anywhere else, and the alternative to a passphrase is writing secrets in the clear.
 */
class BoardSyncRunner(
    private val context: Context,
    private val transport: BoardTransport,
    private val passphrase: String? = null
) {
    private val db = AppDatabase.getInstance(context)

    suspend fun run(
        hasSyncedBefore: Boolean,
        adoptsOnFirstSync: Boolean
    ): BoardSyncOutcome = withContext(Dispatchers.IO) {
        val tasks = db.taskDao().getAllForSync()
        val reminders = db.reminderDao().getAllForSync()
        val storage = db.storageItemDao().getAllForSync()
        val credentials = db.credentialDao().getAllForSync()
        val sketchFolders = SketchStorage.listAllForSync(context)

        // Hashed once, up front. Every later question — what to upload, what a strip's attachments
        // are called, where an arriving hash already lives — is answered from this one pass rather
        // than by reading the same files again.
        val pathsByHash = mutableMapOf<String, String>()
        fun hashed(file: File): String =
            SyncFileStore.hash(file).also { pathsByHash.putIfAbsent(it, file.path) }

        val localIdBySyncId = tasks.associate { it.syncId to it.id }
        val syncIdByLocalId = tasks.associate { it.id to it.syncId }
        val sketchSyncIdByFolder = sketchFolders.associate {
            it.name to SketchStorage.getOrCreateSyncId(it)
        }
        val sketchFolderBySyncId = sketchSyncIdByFolder.entries.associate { it.value to it.key }

        val taskRecords = tasks.map { task ->
            BoardRecords.toRecord(
                task,
                attachments = BoardFiles.attachments(task, ::hashed),
                syncIdOfTask = { syncIdByLocalId[it] },
                syncIdOfSketch = { sketchSyncIdByFolder[it] }
            )
        }
        val storageRecords = storage.map { item ->
            val file = File(item.path)
            BoardRecords.toRecord(item, hash = if (file.exists()) hashed(file) else "")
        }
        val sketchRecords = sketchFolders.map { folder ->
            SyncSketchRecord(
                id = sketchSyncIdByFolder.getValue(folder.name),
                updatedAt = if (SketchStorage.isDeleted(folder)) {
                    SketchStorage.getDeletedAt(folder)
                } else {
                    SketchStorage.lastModified(folder)
                },
                isDeleted = SketchStorage.isDeleted(folder),
                name = SketchStorage.getName(folder).orEmpty(),
                pages = SketchStorage.listPages(folder).map(::hashed),
                createdAt = SketchStorage.getCreatedAt(folder)
            )
        }

        val local = BoardSnapshot(
            tasks = taskRecords,
            reminders = reminders.map(BoardRecords::toRecord),
            storage = storageRecords,
            credentials = credentials.map { BoardRecords.toRecord(it, portablePassword(it)) },
            sketches = sketchRecords
        )

        val outcome = BoardSync(transport).run(
            local = local,
            localHashes = pathsByHash.keys.toSet(),
            hasSyncedBefore = hasSyncedBefore,
            adoptsOnFirstSync = adoptsOnFirstSync,
            readFile = { hash -> pathsByHash[hash]?.let(::File) },
            writeFile = { hash, downloaded ->
                // Landed under this device's own naming, then remembered by hash so the strips
                // applied below can find it. Where it goes depends on nothing but the kind, since
                // a file's name says nothing about which strip wanted it.
                val landed = MediaStorage.newImageFile(context).parentFile
                    ?.let { File(it, hash) }
                    ?: File(context.filesDir, hash)
                downloaded.copyTo(landed, overwrite = true)
                pathsByHash[hash] = landed.path
            }
        )

        if (outcome.failure == null) {
            applyLocally(local, outcome, localIdBySyncId, pathsByHash, sketchFolderBySyncId)
        }
        outcome
    }

    /** Only what actually differs is written — see SyncBoardPlan.plan, which is what decides that.
     * A sync that changed nothing must not touch a single row, or every sync wakes every screen. */
    private suspend fun applyLocally(
        local: BoardSnapshot,
        outcome: BoardSyncOutcome,
        localIdBySyncId: Map<String, Long>,
        pathsByHash: Map<String, String>,
        sketchFolderBySyncId: Map<String, String>
    ) {
        val taskDao = db.taskDao()
        val existingTasks = taskDao.getAllForSync().associateBy { it.syncId }
        // Planned over records on both sides, never records against rows: the plan asks whether
        // two things are equal, and an entity and a record are never equal to each other however
        // faithfully one was built from the other.
        val plan = SyncBoardPlan.plan(
            local.tasks.associateBy { it.id }, outcome.merged.tasks, { it.id }, { it.isDeleted }
        )

        (plan.insert + plan.update).forEach { record ->
            val existing = existingTasks[record.id]
            val applied = BoardRecords.applyTo(
                existing,
                record,
                localIdOfSyncId = { localIdBySyncId[it] },
                localSketchIdOf = { sketchFolderBySyncId[it] }
            )
            val withFiles = BoardFiles.applyAttachments(applied, record) { pathsByHash[it] }
            if (existing == null) taskDao.insert(withFiles) else taskDao.update(withFiles)
        }
        // A tombstone that arrived is written as a tombstone, not removed: this device has to be
        // able to tell the *next* device about the delete too.
        plan.delete.forEach { syncId ->
            existingTasks[syncId]?.let { taskDao.update(it.copy(isDeleted = true)) }
        }

        val reminderDao = db.reminderDao()
        val existingReminders = reminderDao.getAllForSync().associateBy { it.syncId }
        val reminderPlan = SyncBoardPlan.plan(
            local.reminders.associateBy { it.id }, outcome.merged.reminders, { it.id }, { it.isDeleted }
        )
        (reminderPlan.insert + reminderPlan.update).forEach { record ->
            val existing = existingReminders[record.id]
            val applied = BoardRecords.applyTo(existing, record)
            if (existing == null) reminderDao.insert(applied) else reminderDao.update(applied)
        }
        reminderPlan.delete.forEach { syncId ->
            existingReminders[syncId]?.let { reminderDao.update(it.copy(isDeleted = true)) }
        }

        applyStorage(local, outcome, pathsByHash)
        applyCredentials(local, outcome)
        applySketches(local, outcome, pathsByHash)
    }

    /**
     * A library item is only worth a row once its bytes are somewhere.
     *
     * An item whose file hasn't landed yet is skipped rather than written with an empty path — the
     * library would list something that opens nothing, and the row would then look up to date, so
     * no later sync would fix it. Skipped, it simply arrives on the sync after the bytes do.
     */
    private suspend fun applyStorage(
        local: BoardSnapshot,
        outcome: BoardSyncOutcome,
        pathsByHash: Map<String, String>
    ) {
        val dao = db.storageItemDao()
        val existing = dao.getAllForSync().associateBy { it.syncId }
        val plan = SyncBoardPlan.plan(
            local.storage.associateBy { it.id }, outcome.merged.storage, { it.id }, { it.isDeleted }
        )

        (plan.insert + plan.update).forEach { record ->
            val here = existing[record.id]
            val path = pathsByHash[record.hash]
            if (here == null && path == null) return@forEach
            val applied = BoardRecords.applyTo(here, record, pathForNew = path.orEmpty())
            if (here == null) dao.insert(applied) else dao.update(applied)
        }
        plan.delete.forEach { syncId ->
            existing[syncId]?.let { dao.update(it.copy(isDeleted = true)) }
        }
    }

    /**
     * A credential's password has to be re-encrypted before it can be stored.
     *
     * The record carries it under the user's passphrase, which is the only form that can cross
     * between devices; this device keeps it under its own keystore. With no passphrase — or a
     * record that carried no secret — the password already stored is left exactly as it is. Losing
     * a password to a sync that couldn't read it would be losing data; keeping the old one never
     * is, and the title and username still arrive.
     */
    private suspend fun applyCredentials(local: BoardSnapshot, outcome: BoardSyncOutcome) {
        val dao = db.credentialDao()
        val existing = dao.getAllForSync().associateBy { it.syncId }
        val plan = SyncBoardPlan.plan(
            local.credentials.associateBy { it.id }, outcome.merged.credentials,
            { it.id }, { it.isDeleted }
        )

        (plan.insert + plan.update).forEach { record ->
            val here = existing[record.id]
            val applied = BoardRecords.applyTo(here, record, encryptedPassword = storedPassword(record))
            if (here == null) dao.insert(applied) else dao.update(applied)
        }
        plan.delete.forEach { syncId ->
            existing[syncId]?.let { dao.update(it.copy(isDeleted = true)) }
        }
    }

    private fun storedPassword(record: SyncCredentialRecord): String? {
        val secret = passphrase ?: return null
        if (!record.hasPassword) return null
        return runCatching {
            val plain = com.saud.taskstrip.backup.BackupCrypto.decrypt(
                com.saud.taskstrip.backup.BackupCrypto.Encrypted(
                    record.passwordSalt.orEmpty(), record.passwordIv.orEmpty(),
                    record.passwordCipher.orEmpty()
                ),
                secret
            ) ?: return null
            com.saud.taskstrip.security.CredentialCrypto.encrypt(plain)
        }.getOrNull()
    }

    /**
     * A sketch's pages are written back in the record's order, not the order they arrived.
     *
     * Page numbering is a local detail — page1.png, page2.png — and the record is what says which
     * page is which. A note whose pages haven't all landed is left alone entirely rather than
     * written half-finished: half a drawing is worse than yesterday's whole one, and the next sync
     * writes it properly once the rest of the bytes are here.
     */
    private fun applySketches(
        local: BoardSnapshot,
        outcome: BoardSyncOutcome,
        pathsByHash: Map<String, String>
    ) {
        val here = local.sketches.associateBy { it.id }
        val folders = SketchStorage.listAllForSync(context).associateBy {
            SketchStorage.getOrCreateSyncId(it)
        }

        outcome.merged.sketches.forEach { record ->
            if (record == here[record.id]) return@forEach
            val folder = folders[record.id] ?: SketchStorage.noteRef(context, "note_${record.createdAt}")

            if (record.isDeleted) {
                if (SketchStorage.listPages(folder).isNotEmpty()) SketchStorage.deleteNote(folder)
                SketchStorage.setSyncId(folder, record.id)
                return@forEach
            }

            val landed = record.pages.map { pathsByHash[it] }
            if (landed.any { it == null }) return@forEach

            folder.mkdirs()
            SketchStorage.setSyncId(folder, record.id)
            SketchStorage.listPages(folder).forEach { it.delete() }
            landed.filterNotNull().forEachIndexed { index, path ->
                File(path).copyTo(File(folder, "page${index + 1}.png"), overwrite = true)
            }
            if (record.name.isNotBlank()) SketchStorage.setName(folder, record.name)
        }
    }

    /**
     * A credential's password in a form that can leave this device, or nothing.
     *
     * The stored column is keystore-tied and means nothing anywhere else, so it is decrypted here
     * and re-encrypted under the user's passphrase — exactly what BackupHelper does. With no
     * passphrase the password simply doesn't travel; it is never written in the clear.
     */
    private fun portablePassword(
        credential: com.saud.taskstrip.data.CredentialEntity
    ): BoardRecords.PortablePassword? {
        val secret = passphrase ?: return null
        if (credential.encryptedPassword.isBlank()) return null
        return runCatching {
            val plain = com.saud.taskstrip.security.CredentialCrypto.decrypt(credential.encryptedPassword)
            val encrypted = com.saud.taskstrip.backup.BackupCrypto.encrypt(plain, secret)
            BoardRecords.PortablePassword(encrypted.salt, encrypted.iv, encrypted.cipher)
        }.getOrNull()
    }
}
